Bug fixes resulting in the library working reliably

desowin opened the pull request:
These are just the bug fixes extracted from https://github.com/matwey/libopenvizsla/pull/6

desowin committed:
Configure MPSSE according to FTDI Application Note

Enable RTS/CTS flow control as per AN_135.

Set the mask to 0 instead of SIO_SET_BITMODE_REQUEST (0x0b) when
enabling MPSSE. AN_135 tells to set mask to 0 and does not explain what
happens if the mask is something different.

desowin committed:
Fix firmware send buffer race condition

Keep switching between two buffers when reversing the data. The buffer
must not be modified when a transfer is in progress.

desowin committed:
Fix race condition in cha_loop_transfer_callback()

Process read data before resubmitting the libusb transfer. This prevents
new libusb read from overwriting the buffer while frame_decoder_proc() is
accessing it.

desowin committed:
Fix retry decrement logic

Do not decrement try counter past 0. Prior to this change if the loops timed
out the try counter value would be -1 and thus the error path would not be
reached.

matwey commented on src/bit.c:
```diff
@@ -183,7 +183,7 @@ int bit_load_firmware(struct bit* bit, struct cha* cha, struct chb* chb) {
 
 	try = 3;
-	while (try--
+	while (try && try--
```
Nice catch. But why just don't use `--try` instead of `try--` (and optionally increment number of trials in the previous line?)

desowin replied:
As you noted that would require the increment of the number of trial. We are doing 3 retries and thus I retain the "3" intact.

The two choices most likely don't matter at all in optimized compiled binary code. I think this is more readable to humans (if you disagree, I can change it).

matwey replied:
I know. :-) This should be rewritten as

```c
for (try = 3;
	try && (ret = ftdi_write_data(&cha->ftdi, init_cycles, sizeof(init_cycles))) > 0
	&& (ret = chb_get_high(chb, &status)) == 0
	&& !(status & PORTB_DONE_BIT);
	--try);
```

desowin replied:
Rewritten as suggested.

matwey commented on src/chb.c:
```diff
@@ -185,7 +185,7 @@ int chb_switch_program_mode(struct chb* chb) {
 
 	try = 3;
-	while (try-- && (ret = chb_get_high(chb, &status)) == 0 && (status & PORTB_DONE_BIT))
+	while (try && try-- && (ret = chb_get_high(chb, &status)) == 0 && (status & PORTB_DONE_BIT))
```
Same question here.

desowin replied:
Same answer.

matwey replied:
```c
for (try = 3; try && (ret = chb_get_high(chb, &status)) == 0 && (status & PORTB_DONE_BIT); --try) {
	usleep(10000);
}
```

matwey commented on src/chb.c:
```diff
@@ -66,17 +66,43 @@ int chb_open(struct chb* chb) {
+	 * Configure FTDI MPSSE according to AN_135 "FTDI MPSSE Basics"
+	 * 4.2 Configure FTDI Port For MPSSE Use
+	 */
+
```
Extra empty line?

desowin replied:
The comment refers to the code up to the "if (ftdi_set_bitmode(&chb->ftdi, 0, BITMODE_MPSSE) < 0) {". This is just my taste to separate the comments applying to the following multiple code chunks. I can remove it if you insist.

desowin replied:
The comment refers to mutliple code chunks following afterwards (up to the "if (ftdi_set_bitmode(&chb->ftdi, 0, BITMODE_MPSSE) < 0) {"). This is just my taste to have an extra line of separation if it applies to not only the immediate chunk. I can remove this empty line if you insist.

matwey commented on src/chb.c:
```diff
@@ -21,6 +23,14 @@
 #define PORTB_M0_BIT   (1 << 6)
 #define PORTB_M1_BIT   (1 << 7)
 
+static void delay_ms(int ms) {
```
Can we use `msleep` name here for consistency? And `unsigned int` type...

desowin replied:
msleep() is not defined by POSIX. I have updated the code to be portable (unfortunately POSIX first deprecated and then removed the usleep() in favor of nanosleep()).

matwey replied:
I mean we could use `msleep` name instead of `delay_ms`.

desowin replied:
Sorry, misunderstood! Renamed.

matwey commented on src/bit.c:
```diff
@@ -159,16 +160,18 @@ int bit_load_firmware(struct bit* bit, struct cha* cha, struct chb* chb) {
 		}
 
-		if (!(tc = ftdi_write_data_submit(&cha->ftdi, buf, i))) {
+		if (!(tc = ftdi_write_data_submit(&cha->ftdi, buf[bidx], i))) {
```
Hello,

Could you please provide link to the docs? I am not sure why this should be an issue here for single-threaded application.

desowin replied:
I am afraid but it is so low level that I am not aware of any single doc line explaining the reason. Thus, we have to dig deep into the code:
- ftdi_write_data_submit() calls libusb_submit_transfer() under the hood
- libusb_submit_transfer() calls one of its backend function. On Windows the execution eventually ends up in winusbx_submit_bulk_transfer()
- winusbx_submit_bulk_transfer() uses WritePipe with the OVERLAPPED structure
- control returns

Notice that the data was never copied and it is the original buffer that's submitted to the driver. Thus the driver will directly read from the provided buffer (possibly locking it into memory; locking means that the data won't get swapped, the initiating application can still modify it). On Windows, kernel code executes at PASSIVE_LEVEL or higher. Driver code operating at PASSIVE_LEVEL can basically interrupt any user application at any time. What happens here, is the driver code accesses the the buffer while we update it.

Thus we have to either use two buffers (as done here), or simply use synchronous ftdi_write_data().

I do not consider this to be a Windows only issue. This basically can happen in any system that does not copy the buffer.

matwey replied:
Hm... Now it is getting clear to me. Thank you. I get used that in Linux the data is copied to the kernel at the syscall in 99% cases.

matwey replied:
Can we rephrase the patch as the following: `int bidx` -> `size_t bidx` (since bidx is used as array index variable)?
And redo `while(length) {` loop as the following
```
for (bidx = 0; length; bidx = (bidx + 1) % 2) {
```

desowin replied:
I have applied the `size_t` and `while` -> `for` change mentioned in above comment.

matwey commented on src/cha.c:
```diff
@@ -474,6 +466,14 @@ static void cha_loop_transfer_callback(struct libusb_transfer* transfer) {
+
+				loop->break_loop = 1;
+			}
+
```
It is not what initially supposed to be here. We should parse while the request for the following data chunk is already sent.

desowin replied:
The same explanation applies as in the write case, except now the data flow is opposite. The driver can write the data (and it is trivial to experience on Windows on my computer) while we haven't yet parsed the data that was in the buffer.

matwey replied:
What do you think? I would like to introduce second `libusb_transfer` object here in order not to wait for parser done before firing next data request to the hardware.

desowin replied:
I think it might be worthwhile to do so alongside some refactor (that is outside the scope of bug fix).

If you look at current extcap implementation it calls `ov_capture_dispatch()` with `count` equal 1 - and thus the second transfer here won't help. This is because Wireshark lets extcap know that it should stop by closing the pipe. If there was mechanism to let the user callback to notify library whether it should continue or not (possibly instead of the count) then it would make sense in my opinion to have two transfers here.

matwey replied:
Actually, there is the mechanism. `ov_capture_dispatch` should properly handle SIGTERM and co. This is how pcap modules work: https://github.com/matwey/libpcap/commit/48341c7dce3ad3ea0ec369d4e2d1e6454d23f352

desowin replied:
I am probably missing something, but I can't see anything related to SIGTERM in the linked commit. Neither I see anything in the library except the sample that has SIGTERM handler.

Am I unerstanding it correctly to simply `raise(SIGTERM);` inside the callback?

matwey replied:
In Linux, signals are received when application makes a syscall. Then the syscall is returned with special error code and registered handler is executed before that.

So, if `LIBUSB_ERROR_INTERRUPTED` is returned from `libusb_handle_events_completed`, then it means that we are received a signal, so the function is terminated with special error code -2 in order to show the user that the loop has been interrupted by a signal.

matwey replied:
I think that I am going to merge this PR and we will recover the performance later.

desowin committed:
Add delays when switching into programming mode

Make sure the pins are in the desired state for atleast 1 ms. This prevents
the FPGA from missing/filtering out the short pulse.

matwey commented:
Thanks
